File: Baekjoon/1932_정수 삼각형/B_1932_2.js
const fs = require('fs')
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n')

const n = parseInt(input[0])
let triangle = []

for (let i = 1; i <= n; i++) {
    const nums = input[i].split(' ').map(v => Number(v))
    triangle.push(nums)
}

for (let i = 1; i < n; i++) {
    const leng = triangle[i].length
    for (let j = 0; j < leng; j++) {
        if (j === 0) {
            triangle[i][j] += triangle[i - 1][j]
        } else if (j === leng - 1) {
            triangle[i][j] += triangle[i - 1][j - 1]
        } else {
            triangle[i][j] += Math.max(triangle[i - 1][j - 1], triangle[i - 1][j])
        }
    }
}

console.log(Math.max(...triangle[n - 1]))

File: Baekjoon/1932_정수 삼각형/B_1932_2.py
import sys


n = int(sys.stdin.readline())
triangles = []
dp = []
for i in range(n):
    triangle = list(map(int, sys.stdin.readline().split()))
    triangles.append(triangle)
    dp_item = [0] * (i + 1)
    dp.append(dp_item)

dp[0][0] = triangles[0][0]

for i in range(1, n):
    for j in range(i + 1):
        if j == 0:
            dp[i][j] = triangles[i][j] + dp[i - 1][j]
        elif j == i:
            dp[i][j] = triangles[i][j] + dp[i - 1][j - 1]
        else:
            dp[i][j] = triangles[i][j] + max(dp[i - 1][j - 1], dp[i - 1][j])

sys.stdout.write(str(max(dp[n - 1])))
